### PR TITLE
Only depend on ipaddress in python2

### PR DESCRIPTION
**Pull request by asottile**
The ipaddress module became a standard module in python3.3.

This uses [environment markers][pep508] to only select the `ipaddress` module
in python2.7.  In setuptools, the most compatible way to accomplish this is
through the `extras_require` field (as suggested in the [wheel][wheel] docs).

[pep508]: https://www.python.org/dev/peps/pep-0508/#id23
[wheel]: https://wheel.readthedocs.io/en/latest/#defining-conditional-dependencies

### CODE REVIEW

**Comment by asottile** [comment]
CC @yliaog @roycaihw -- would love a review :)

File: setup.py [hunk 1]
```diff
@@ -27,8 +27,18 @@
...
+REQUIRES = []
 with open('requirements.txt') as f:
-    REQUIRES = f.readlines()
+    for line in f:
+        line, _, _ = line.partition('#')
+        line = line.strip()
+        if ';' in line:
+            requirement, _, specifier = line.partition(';')
+            for_specifier = EXTRAS.setdefault(':{}'.format(specifier), [])
+            for_specifier.append(requirement)
```
Comment (yliaog) [inline setup.py]: what is the purpose of "for_specifier.append(requirement)" but for_specifier is not kept anywhere?

**Comment by yliaog** [comment]
could you please write a test case for it?

**Comment by asottile** [comment]
I'm not sure how I would test this, it's really only a quality of the installation itself -- generally `setup.py` isn't really testable (either it installs or it doesn't!).  tox covers this behaviour (it invokes the installation).

I've noted in `travis-ci` that the python2.7 is still installing `ipaddress` and the py35 / py36 installs are not -- meaning it works as intended.

**Comment by yliaog** [comment]
would you be able to add a line in test-requirements.txt to test it?

File: setup.py (hunk 1 above)
Comment (asottile) [reply setup.py]: `for_specifier` is the list inside `EXTRAS[f":{spec}"]` -- I'd use `collections.defaultdict(list)` here but setuptools requires specific types here.

**Comment by asottile** [comment]
> would you be able to add a line in test-requirements.txt to test it?

the `setup.py` is invoked by `tox` already, so no need to add anything here

File: setup.py (hunk 1 above)
Comment (asottile) [reply setup.py]: The trick here is [`setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault)

**Comment by yliaog** [comment]
i think it's possible to write a test case to check which python version is run, and if python version is 2.7, then check if ipaddress module is installed, and also where it is installed, etc.

But i agree manually checking that it works is probably good enough.

**Comment by asottile** [comment]
Presumably the tests would crash on an `ImportError` for `ipaddress` if it weren't being installed.  I think that's covered :)

**Pull request merged**